File: misc/leetcode/lowest-common-ancestor-of-a-binary-tree.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """

        def walk(root, p, q):
            if root is None:
                return None, 0

            c = 0
            if root == p or root == q:
                c += 1
            t0, c0 = walk(root.left, p, q)
            if c0 == 2:
                return t0, c0
            t1, c1 = walk(root.right, p, q)
            if c1 == 2:
                return t1, c1
            c += c0 + c1
            return root, c

        ans, _ = walk(root, p, q)
        return ans

File: misc/leetcode/lowest-common-ancestor-of-a-binary-tree.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <cstdio>
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
};

class Solution {
   public:
    TreeNode* find(TreeNode* root, TreeNode* p, TreeNode* q, int* covers) {
        if (root == NULL) return NULL;

        int c = 0;
        if ((root == p) || (root == q)) c += 1;

        int c0 = 0, c1 = 0;
        TreeNode* t0 = find(root->left, p, q, &c0);
        if (c0 == 2) {
            *covers = 2;
            return t0;
        }
        TreeNode* t1 = find(root->right, p, q, &c1);
        if (c1 == 2) {
            *covers = 2;
            return t1;
        }
        c += (c0 + c1);
        *covers = c;
        return root;
    }

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int covers = 0;
        TreeNode* t = find(root, p, q, &covers);
        return t;
    }
};
